src/libraries/System.Net.Primitives/src/System/Net/IPAddress.cs

```diff
@@ -188,19 +179,43 @@ public IPAddress(ReadOnlySpan<byte> address)
             }
             else if (address.Length == IPAddressParserStatics.IPv6AddressBytes)
             {
-                _numbers = new ushort[NumberOfLabels];
-
-                for (int i = 0; i < NumberOfLabels; i++)
-                {
-                    _numbers[i] = (ushort)(address[i * 2] * 256 + address[i * 2 + 1]);
-                }
+                _numbers = ReadUInt16NumbersFromBytes(address);
             }
             else
             {
                 throw new ArgumentException(SR.dns_bad_ip_address, nameof(address));
             }
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static ushort[] ReadUInt16NumbersFromBytes(ReadOnlySpan<byte> address)
+        {
+            if (!BitConverter.IsLittleEndian)
+            {
+                return MemoryMarshal.Cast<byte, ushort>(address).ToArray();
+            }
+
+            if (Vector128.IsHardwareAccelerated)
+            {
+                Vector128<byte> bytes = Vector128.Shuffle(
```

Comment:
I noticed a potential problem with `Vector128.IsHardwareAccelerated` - so currently it returns `true` for SSE1 or SSE2.
So e.g. the code you wrote actually relies on at least SSSE3 to present, otherwise `Vector128.Shuffle` fallbacks to extremely slow software implementation. SSE2 is a baseline for Crossgen and NativeAOT, e.g. here is what NativeAOT prints: 
![image](https://user-images.githubusercontent.com/523221/191162223-3e79c2db-ac82-4e8f-a09e-1ce3ad0bd2fa.png)

(only half of the codegen on the screen) cc @tannergooding @MichalStrehovsky

Reply:
So for explicit checks NativeAOT (Presumably, crossgen too) generates run-time checks, e.g.
![image](https://user-images.githubusercontent.com/523221/191163632-d01b5c06-063c-4f14-a73c-93332fdca865.png)

but it doesn't generate it for `Vector128.Shuffle` and emits raw software impl

Reply:
A bit similar to https://github.com/dotnet/runtime/issues/75770 where without perf-investigation astonishing results can happen.

Reply:
(Why are there calls to Debug.Fail showing up in that disasm?)

Reply:
> (Why are there calls to Debug.Fail showing up in that disasm? Is that checked rather than release?)

yep, but even in Release it will be a slow loop-based impl

Reply:
> I noticed a potential problem with Vector128.IsHardwareAccelerated - so currently it returns true for SSE1 or SSE2.

This doesn't sound right. It should only be true for `SSE2+`, much like `Vector<T>`.

>  SSE2 is a baseline for Crossgen and NativeAOT

I believe crossgen is actually SSE4.1 today (a 13-15yo baseline) (CC. @davidwrighton).

> Vector128.Shuffle fallbacks to extremely slow software implementation

Yes, `byte` shuffling was only added to x86/x64 in SSSE3 (~2006) and we don't currently have the advanced logic to detect a case like this and replace it with a more efficient sequence for SSE2 (baseline ISA) only.

It's worth noting in this case we could do: a "rotate" instead. We're going from `0, 1, 2, 3` to `1, 0, 3, 2`. Which is just `(x << 8) | (y >> 8)` at a `word-wise` basis.

----------------

More generally this is one of those "issues" with AOT and having a 20 yo baseline. We are increasingly unlikely to encounter such a machine as there is no hardware in any major cloud provider this old and even something like [Steam Hardware Survey](https://store.steampowered.com/hwsurvey) (see Other Settings) tracks 99.15% of reporting hardware having SSE4.1 support. SSE4.1 is so old even Windows/OSX on Arm64 will emulate the ISAs when transpiling x86/x64 apps.

Its potentially worth considering "upping" the default (from the current `x86-x64-v1` to `x86-x64-v2` instead) and giving users a diagnostic that we aren't portable to some small subset of machinery without them explicitly opting in.

Reply:
@tannergooding I agree that the baseline should probably be higher, because even when arm64 CPUs emulate x64 they support up to 4.1 typically.
But currently the baseline is SSE2, you can check it by executing `dotnet publish -c Release -r win-x64 /p:PublishAOT=true -v:d` and find the exact command passed to LLC - no `--instruction-set` is passed. You can also copy the whole command and play with `-codegenopt:JitDisasm=Main` to see that for e.g. Ssse3.IsSupported it emits a run-time check

Reply:
Right, I'm just saying this is a very AOT specific consideration and something that we should look at more in depth for .NET 8.

I expect targeting SSE2 as a baseline is unrealistic for a majority of our customers/community. It also pessimizes core scenarios like AOT compared to the jitted counterparts and causes x64 to be much worse than Arm64.

If we push for upping the baseline, then this becomes "moot" and we no longer have to consider these edge cases *and* we can also simplify a ton of logic in the JIT/various parts of the libraries.

Reply:
> It's worth noting in this case we could do: a "rotate" instead.

I changed it to use a rotate instead of a shuffle.

Reply:
and I logged https://github.com/dotnet/designs/pull/272 to track potentially changing the baseline.

there are other options as well, such as changing `IsHardwareAccelerated` to only be `true` on SSE4.1+
but still allowing partial acceleration otherwise.